server/tilesource/base.py

```diff
@@ -54,6 +59,104 @@ def __init__(self):
         self.sizeX = None
         self.sizeY = None
 
+    def _calculateWidthHeight(self, width, height, sourceWidth, sourceHeight):
+        """
+        Given a source width and height and a maximum destination width and/or
+        height, calculate a destination width and height that preserves the
+        aspect ratio of the source.
+
+        :param width: the destination width.  None to only use height.
+        :param height: the destination height.  None to only use width.
+        :param sourceWidth: the width of the source data.
+        :param sourceHeight: the height of the source data.
+        :returns: the width and height that is no larger than that specified
+                  and preserves aspect ratio.
+        """
+        if sourceWidth == 0 or sourceHeight == 0:
+            return 0, 0
+        # Constrain the maximum size if both width and height weren't
+        # specified, in case the image is very short or very narrow.
+        if height and not width:
+            width = height * 16
```

Comment:
I know this is existing code, but what is the worst that happens if we didn't effectively bound ``width <= height * 16``? If the image is naturally very short or narrow, shouldn't we just return it as-is?